### ISSUE TITLE
Type incompatibility in multi-venv project using PEP-561 stubs

### ISSUE THREAD

**Issue by MHDante** [microsoft/pyright#1508]
I recognize this edge case is quite esoteric. but here's the bug report anyway.

**Describe the bug**
Incompatible types between library code that uses the same package.

**To Reproduce**

Create a workspace with the following `pyrightconfig.json`:
```
{
    "venvPath": "venvs",
    "typeCheckingMode": "strict",
    "pythonVersion": "3.7",
    "exclude": [
        "venvs"
    ],
    "useLibraryCodeForTypes": true,
    "executionEnvironments": [
        {
            "root": "shared_lib/src"
        },
        {
            "root": "root1/src",
            "venv": "venv1",
            "extraPaths": ["shared_lib/src"]
        },
        {
            "root": "root2/src",
            "venv": "venv2",
            "extraPaths": ["shared_lib/src"]
        }
    ]
}
```

2. Create the venvs for each root
3. Import a  PEP-561 compliant stub package
     - In this case, it was [`grpc-stubs`](https://github.com/shabbyrobe/grpc-stubs)
4. access a typed signature across the library and see errors

**Expected behavior**
No errors.

**Screenshots or Code**

I've uploaded a minimal example workspace to  https://github.com/MHDante/pyright-error

here's the commandline output:
```
PS C:\Repos\pyright-error> pyright .
Loading configuration file at C:\Repos\pyright-error\pyrightconfig.json
Assuming Python platform Windows
stubPath C:\Repos\pyright-error\typings is not a valid directory.
Searching for source files
Found 5 source files
C:\Repos\pyright-error\root1\src\test1.py
  10:33 - error: Argument of type "Server" cannot be assigned to parameter "server" of type "Server" in function "start_insecure"
    "Server" is incompatible with "Server" (reportGeneralTypeIssues)
C:\Repos\pyright-error\root2\src\test2.py
  10:33 - error: Argument of type "Server" cannot be assigned to parameter "server" of type "Server" in function "start_insecure"
    "Server" is incompatible with "Server" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 infos
Completed in 0.844sec
PS C:\Repos\pyright-error>
```
Here's the vscode error:

![image](https://user-images.githubusercontent.com/4652122/108512269-f833d100-728e-11eb-86b7-84192907dcf8.png)

**VS Code extension or command-line**

both.

CommandLine: `pyright 1.1.112`
vscode: `pylance 2021.2.3`

**Additional context**
Incidentally, this doesn't happen if I auto-generate the stubs (in the typings directory) 

If this is not a bug, (as it could be that the stubs are incompatible). Is there a way to coalesce types? or is there a way to install a PEP-561 package into the typings directory?

Once again, you guys are super cool.

**Comment by erictraut** [comment]
Thanks for the great repro steps. 

When you say "multi-root workspace", are you referring to the multiple executionEnvironment roots you've defined in pyrightconfig.json, or are you referring to what VS Code calls a multi-root workspace (i.e. the "Add file to workspace..." menu item)? I don't see any VS Code workspace definition file in your example, so I'm not sure what you mean.

I tried following your instructions precisely, but I'm not able to repro the problem. It seems to work fine for me.

<img width="454" alt="Screen Shot 2021-02-19 at 3 40 35 PM" src="https://user-images.githubusercontent.com/7040122/108569289-de16e480-72c8-11eb-97de-09c7b50112bb.png">

**Comment by MHDante** [comment]
Yeah. I think I am somewhat confused about the terminology. I must have heard this used elsewhere and I am using it incorrectly. The real issue lies in the search path used by the different folders.

This can be reproduced without VS code as follows:

assumes `git`, `pyright` and `python`  (3.7.9) are installed.


In Powershell:
``` 
git clone https://github.com/MHDante/pyright-error
cd .\pyright-error\
cd .\venvs\
python -m venv venv1
python -m venv venv2
cd ..\root1\
..\venvs\venv1\Scripts\activate
pip install -r .\requirements.txt
deactivate
cd ..\root2\
..\venvs\venv2\Scripts\activate
pip install -r .\requirements.txt
deactivate
cd ..
pyright .
```

In bash:

``` 
git clone https://github.com/MHDante/pyright-error
cd ./pyright-error/
cd ./venvs/
python -m venv venv1
python -m venv venv2
cd ../root1/
source ../venvs/venv1/bin/activate
pip install wheel
pip install -r ./requirements.txt
deactivate
cd ../root2/
source ../venvs/venv2/bin/activate
pip install wheel
pip install -r ./requirements.txt
deactivate
cd ..
pyright .
```

**Comment by erictraut** [comment]
The problem is that you have not specified which venv should be used when type checking the source files in the `shared_lib/src` directory.

You have two choices:
1. If you want to type check these files, update the first first of the three execution environments to include `"venv": "venv1"`.
2. If you don't want to type check these files, add that directory to the "exclude" section of pyrightconfig.json and delete the first of the three execution environments.

**Comment by erictraut** [comment]
I'm going to close the bug, but if I've misdiagnosed it or you still have questions, feel free to reopen.

**Comment by MHDante** [comment]
I understand and have verified this behavior in the project that led me here. Apologies for wasting your time.

**Comment by MHDante** [comment]
See https://github.com/microsoft/pyright/issues/1508#issuecomment-782811645 for context

**Comment by MHDante** [comment]
For anyone else who ends up here, I've added a template with this behaviour (working as expected)

https://github.com/MHDante/pyright-test/releases/tag/1508_fixed

to verify, run:

```
git clone --branch 1508_fixed https://github.com/MHDante/pyright-test.git
cd pyright-test/
chmod +x setup.sh
./setup.sh 
pyright .
```

**Comment by MHDante** [comment]
Since this is still happening due to the problems in #1523, I'd like to suggest reopening the issue. 

I can also suggest an interim solution:

Perhaps, it would be useful to add a new error category for the incompatible type warning when both types have the same name (and perhaps relative path in their respective venvs?). This way, we can reduce the severity of the error. 

That said, I can see where this check would be useful, especially if different venvs have different package versions.

**Comment by erictraut** [comment]
The problem that is reported in this issue is expected behavior because of the way you've configured the project. If you fix the configuration as I suggested above, then you should no longer see the error message indicated in this report. You may see other unrelated issues as reported in #1523, but we already have a bug report covering those. If you think that there's another issue (beyond the misconfiguration in this one and the bug described in #1523), please open a new bug report with specific steps.

Did you make the configuration change I suggested above, and did that eliminate the problem you reported? 

I don't understand the workaround you suggested above. The problem that I saw when I followed your repro steps did not involve an incompatible type warning. It involved an unresolved import error because a venv was not specified for the `shared_lib/src` execution environment.

**Comment by MHDante** [comment]
I just want to say that I'm extremely thankful for your work and dedication to this project. I apologize for not being clearer with my wording. This issue likely affects a small portion of your users, so I would understand if it is de-prioritized. However, my intention is just to document the error's idiosyncrasies .

I'm pointing out the following error:

```
C:\Repos\pyright-error\root1\src\test1.py
  16:33 - error: Argument of type "Server" cannot be assigned to parameter "server" of type "Server" in function "start_insecure"
    "Server" is incompatible with "Server" (reportGeneralTypeIssues)
```

There's three cases

-----

### Case 1: 

This is the original post I submitted. I originally assumed that I could leave `shared_lib` without a venv (since it is never run directly, it's a shared library). This was labeled a misconfiguration.

When running `pyright .` I get the error mentioned above. (`Argument of type "Server" cannot be assigned to parameter`) I think that the reason you might have gotten the unresolved import error is because you didn't have `grpc` in the system site-packages. (which I failed to mention)

I would argue that the error I got is undesired behavior, since the library does exist (in the system site packages), and is the same one (same version) as in the `venv1` environment.

 - Navigating to the definition of `grpc` in `import grpc` in test1.py goes to the `venv1` site-packages:

![image](https://user-images.githubusercontent.com/4652122/108892278-addb8880-75dd-11eb-82bd-85b1ae127f37.png)

 - Navigating to the definition of  `grpc` in `import grpc` in my_lib.py goes to the system site-packages:
 
![image](https://user-images.githubusercontent.com/4652122/108892950-80430f00-75de-11eb-99b1-dda77411efe3.png)


 - `pyrightconfig.json` :

```
{
    "venvPath": "venvs",
    "typeCheckingMode": "strict",
    "pythonVersion": "3.7",
    "exclude": [
        "venvs"
    ],
    "useLibraryCodeForTypes": true,
    "executionEnvironments": [
        {
            "root": "shared_lib/src"
        },
        {
            "root": "root1/src",
            "venv": "venv1",
            "extraPaths": ["shared_lib/src"]
        },
        {
            "root": "root2/src",
            "venv": "venv2",
            "extraPaths": ["shared_lib/src"]
        }
    ]
}
```

If I run `pyright .` (from the pyright-test folder), I get:
```
C:\Repos\pyright-test\proj1\src\test1.py
  11:9 - error: Method "GetStatus" overrides class "SharedClass" in an incompatible manner
    Parameter 3 type mismatch: base parameter is type "ServicerContext", override parameter is type "ServicerContext"
      "ServicerContext" is incompatible with "ServicerContext" (reportIncompatibleMethodOverride)
  16:33 - error: Argument of type "Server" cannot be assigned to parameter "server" of type "Server" in function "start_insecure"
    "Server" is incompatible with "Server" (reportGeneralTypeIssues)
C:\Repos\pyright-test\proj2\src\test2.py
  11:9 - error: Method "GetStatus" overrides class "SharedClass" in an incompatible manner
    Parameter 3 type mismatch: base parameter is type "ServicerContext", override parameter is type "ServicerContext"
      "ServicerContext" is incompatible with "ServicerContext" (reportIncompatibleMethodOverride)
  19:33 - error: Argument of type "Server" cannot be assigned to parameter "server" of type "Server" in function "start_insecure"
    "Server" is incompatible with "Server" (reportGeneralTypeIssues)
4 errors, 0 warnings, 0 infos
Completed in 0.958sec
```

If I uninstall grpc and its stubs
```
pip uninstall grpcio
pip uninstall grpc-stubs
```
I get this output:

```
PS C:\Repos\pyright-test> pyright .
Loading configuration file at C:\Repos\pyright-test\pyrightconfig.json
Assuming Python platform Windows
stubPath C:\Repos\pyright-test\typings is not a valid directory.
Searching for source files
Found 8 source files
C:\Repos\pyright-test\shared_lib\src\shared_code\my_lib.py
  1:8 - error: Import "grpc" could not be resolved (reportMissingImports)
  3:38 - error: Type of "Server" is unknown (reportUnknownMemberType)
  3:30 - error: Type of parameter "server" is unknown (reportUnknownParameterType)
  4:5 - error: Type of "add_insecure_port" is unknown (reportUnknownMemberType)
  5:5 - error: Type of "start" is unknown (reportUnknownMemberType)
  6:5 - error: Type of "wait_for_termination" is unknown (reportUnknownMemberType)
C:\Repos\pyright-test\shared_lib\src\stubbed_code\shared_class.pyi
  1:8 - error: Import "grpc" could not be resolved (reportMissingImports)
  9:18 - error: Type of "ServicerContext" is unknown (reportUnknownMemberType)
  9:9 - error: Type of parameter "context" is unknown (reportUnknownParameterType)
9 errors, 0 warnings, 0 infos
Completed in 0.929sec
PS C:\Repos\pyright-test> 
```


-----

### Case 2:

This is applying the configuration fix you suggested. This does not produce the error. 🎉 

 - Navigating to the definition of `grpc` in `import grpc` in test1.py goes to the `venv1` site-packages:

![image](https://user-images.githubusercontent.com/4652122/108892278-addb8880-75dd-11eb-82bd-85b1ae127f37.png)


 - Navigating to the definition of  `grpc` in `import grpc` in my_lib.py goes to the  `venv_common` site-packages:
 
![image](https://user-images.githubusercontent.com/4652122/108893370-06f7ec00-75df-11eb-9c00-01de08215b5b.png)


Despite these being different files, you get no error.

```
PS C:\Repos\pyright-test> pyright .
Loading configuration file at C:\Repos\pyright-test\pyrightconfig.json
Assuming Python platform Windows
stubPath C:\Repos\pyright-test\typings is not a valid directory.
Searching for source files
Found 8 source files
0 errors, 0 warnings, 0 infos 
Completed in 0.913sec
```

 - `pyrightconfig.json` :

```
{
    "venvPath": "venvs",
    "typeCheckingMode": "strict",
    "pythonVersion": "3.7",
    "exclude": [
        "venvs"
    ],
    "useLibraryCodeForTypes": true,
    "executionEnvironments": [
        {
            "root": "shared_lib/src",
            "venv": "venv_common"
        },
        {
            "root": "proj1/src",
            "venv": "venv1",
            "extraPaths": ["shared_lib/src"]
        },
        {
            "root": "proj2/src",
            "venv": "venv2",
            "extraPaths": ["shared_lib/src"]
        }
    ]
}
```

-----

### Case 3:

By adding a venv at the top level in `pyrightconfig.json`, _Even while keeping the inner venv_ (venv_common), I get the type incompatibility errors.

This happens even though the symbols resolve to the same files as in Case 2.


 - Navigating to the definition of `grpc` in `import grpc` in test1.py goes to the `venv1` site-packages:

![image](https://user-images.githubusercontent.com/4652122/108892278-addb8880-75dd-11eb-82bd-85b1ae127f37.png)


 - Navigating to the definition of  `grpc` in `import grpc` in my_lib.py goes to the  `venv_common` site-packages:
 
![image](https://user-images.githubusercontent.com/4652122/108893370-06f7ec00-75df-11eb-9c00-01de08215b5b.png)


 - `pyrightconfig.json` :

```
{
    "venvPath": "venvs",
    "typeCheckingMode": "strict",
    "pythonVersion": "3.7",
    "exclude": [
        "venvs"
    ],
    "venv": "venv_common",
    "useLibraryCodeForTypes": true,
    "executionEnvironments": [
        {
            "root": "shared_lib/src",
            "venv": "venv_common"
        },
        {
            "root": "proj1/src",
            "venv": "venv1",
            "extraPaths": ["shared_lib/src"]
        },
        {
            "root": "proj2/src",
            "venv": "venv2",
            "extraPaths": ["shared_lib/src"]
        }
    ]
}
```

-----

### Summary

With that said, It seems like the conditions for type compatibility are either bugged, or unintuitive at best.

In Case 1 (with the packages installed in the system site packages):

`Server` in `C:\Users\mhdcj\AppData\Roaming\Python\Python37\site-packages\grpc-stubs\__init__.pyi`
and 
`Server` in `C:\Repos\pyright-test\venvs\venv1\lib\site-packages\grpc-stubs\__init__.pyi`

are considered **incompatible** 

In Case 2:

`Server` in `C:\Repos\pyright-test\venvs\venv_common\lib\site-packages\grpc-stubs\__init__.pyi`
and 
`Server` in `C:\Repos\pyright-test\venvs\venv1\lib\site-packages\grpc-stubs\__init__.pyi`

are considered **compatible**

In Case 3:


`Server` in `C:\Repos\pyright-test\venvs\venv_common\lib\site-packages\grpc-stubs\__init__.pyi`
and 
`Server` in `C:\Repos\pyright-test\venvs\venv1\lib\site-packages\grpc-stubs\__init__.pyi`

are considered **incompatible**
_despite being the same files as in Case 2!!_

-----

### Sugestion:

My suggestion is that you compare the split the paths to be relative to `site-packages`.
In this case, all the paths above would resolve to: 
`.\grpc-stubs\__init__.pyi`

And we can use that to determine compatibility.

I understand that this could cause a problem if the types are not 100% compatible. (different versions or different packages with the same name)

 So perhaps it can still be an error, but in its own category (maybe `typeMismatchSameName` instead of `reportGeneralTypeIssues`) 

-----

Anyway,  I hope this is helpful, since at the moment I kinda feel like this:
<img src="https://user-images.githubusercontent.com/4652122/108927025-83a1bf00-760d-11eb-84b2-c9bcca1c71e0.png" width="200">

**Comment by MHDante** [comment]
The above post was written with 
pyright 1.1.112 
pylance 2021.2.4-pre.1

I've upgraded to 
pyright 1.1.114
and the behaviour is still the same in all 3 cases.

**Comment by erictraut** [comment]
OK, thanks for the additional details. I'll look into it when I have a chance. I think I'll focus first on #1523 and see what problems that clears up.

**Comment by erictraut** [comment]
After thinking about this more, I think the whole "multi-venv" feature was ill conceived. The right way to handle this is to run separate instances of the analyzer on different projects. We already support this in the VS Code extension for Pyright & Pylance in the form of mult-root workspace support. You would place a pyrightconfig.json file at the root of each workspace, and the VS Code extension creates independent instances of the type checker for each workspace. That's really what's needed here, and we already have that support.

The problem with supporting different venvs within the same type checker instance is that each file tracked by the type checker is analyzed only once. It can't be analyzed multiple times in the context of different venvs. When I added "execution environment" support initially, my assumption was that each execution environment would be distinct from the others with no overlapping source files. But now I'm seeing that people are using it in ways that go far beyond its original use.

So my recommendation is that we remove the ability to specify a separate "venv" for each execution environment. All execution environments specified in the pyrightconfig.json will use the same Python environment for analysis. Each execution environment will still be able to specify different "extraPaths" and different pythonVersion and pythonPlatform. For people who really need different venvs, they can create separate pyrightconfig.json files, one for each root.

@jakebailey, any thoughts here? Do you agree with this approach?

**Comment by jakebailey** [comment]
I'm not sure if I agree entirely. I think multi-root really only works if it really is distinct trees. Multi-root already has troubles with overlapping, and each of those roots may also have their own interpreters set as well. If we are constructing execution environments based on workspaces (I think that's how it works) then I don't know how eliminating `venv` fixes much. `venv` to me is just an alternative to specifying an interpreter path.

The main point of contention for me is this:

> The problem with supporting different venvs within the same type checker instance is that each file tracked by the type checker is analyzed only once. It is not analyzed multiple times in the context of different venvs. That would violate a bunch of assumptions in the code.
>
> When I added "execution environment" support initially, my assumption was that each execution environment would be distinct from the others with no overlapping source files. But now I'm seeing that people are using it in ways that go far beyond its original use.

Ideally, they would be independent but allow overlaps within their own context. Sure, it gets weird for the editor (that's another problem I think we could work out), but I don't see a reason why multiple analyzers wouldn't be able to share files and interpret them according to their needs. I view this as not much different than someone having two multi-root workspaces where typeshed is shared but each analyzer can assume different python versions for each (and therefore do different `sys.platform` checks). It's still resolving code, and typeshed is effectively just another set of search paths like any other.

I also don't like that this will mean that you'd need multiple configs and pyright invocations to do something that was previously working before as one whole unit. I don't want to defend a feature removal like this by saying VS Code workspaces do it, when previously they would have been operating just fine (and still want a CLI to emulate the behavior).

I think we should maybe take a look at how TS does this; the pyright (and pylance) projects themselves are individual TS monorepo codebases that share code effectively, reference each other in configs, etc, without trouble (but, we don't use multi-root as that'd be a nightmare). It'd be interesting to see how they choose to draw the line, where settings stop getting applied, etc. But, I know that they have the benefit of not actually analyzing real JS code and not really needing to do any of the `sys.platform` checks, differing module resolution behaviors, etc...

> So my recommendation is that we remove the ability to specify a separate "venv" for each execution environment. All execution environments specified in the pyrightconfig.json will use the same Python environment for analysis. Each execution environment will still be able to specify different "extraPaths" and different pythonVersion and pythonPlatform. For people who really need different venvs, they can create separate pyrightconfig.json files, one for each root.

With the above, a `venv` to me still effectively specifies a interpreter version and a set of paths. If we eliminate that, all that I feel is happening is that people manually set the same values, but it all gets fed to the same place anyway... So I don't know that he status quo changes.

**Comment by jakebailey** [comment]
I think what may be needed is to sit down and define all of these different scenarios, what people are using them for, how they should interact with multi-root, etc, and see where that leads. But I do think that overlapping is not uncommon, and that we will need to find a way to properly handle them both in type checking as well as in the UI.

**Comment by MHDante** [comment]
Hey, So I just want to raise that I found a  configuration that actually solves all of my use-cases. 
 - Supports package imports that import from themselves using absolute import statements (such as torch)
 - Supports sharing a library type across projects with different venvs (Hard to explain, see above)
 
It's a bit embarrassing, since the solution seems to be what Eric suggested above (but I misunderstood) . I'll add it as Case 4 Below.

I'm also unsure if the reduction in false negatives are changes that came about from recent commits to this repo.

That said, I don't recommend closing the issue. While I have some confusion as to how the multi-venv file resolution works and type matching works. Intuition would likely lead future users into one of the 3 cases I mentioned above. I would argue that all 4 cases should not cause the following issue:

```
C:\Repos\pyright-error\root1\src\test1.py
  16:33 - error: Argument of type "Server" cannot be assigned to parameter "server" of type "Server" in function "start_insecure"
    "Server" is incompatible with "Server" (reportGeneralTypeIssues)
```

Furthermore, this approach requires the creation of 4 venvs for a project that (at runtime) only needs 2 venvs.

Anyways, here is the case.

-----

### Case 4:

Introduce a 4th venv (`venvs/fallback`). This should include common types across all three projects, as well as types that import themselves using absolute imports. We then set this venv as the venv in the top level of `pyrightconfig.json`.

This produces no errors.

Paths (although, at this point, I'm pretty sure they're not relevant to type resolution)

`test1.py` : `C:\Repos\pyright-test\venvs\venv1\lib\site-packages\grpc\__init__.py`
`my_lib.py` : `C:\Repos\pyright-test\venvs\venv_common\lib\site-packages\grpc\__init__.py`

 - `pyrightconfig.json` (thanks for adding support for comments)

```jsonc
{
    "venvPath": "venvs",
    "typeCheckingMode": "strict",
    "pythonVersion": "3.7",
    "exclude": [
        "venvs"
    ],
    // I replace venv_common with fallback. 
    "venv": "fallback",
    "useLibraryCodeForTypes": true,
    "executionEnvironments": [
        {
            "root": "shared_lib/src",
            "venv": "venv_common"
        },
        {
            "root": "proj1/src",
            "venv": "venv1",
            "extraPaths": ["shared_lib/src"]
        },
        {
            "root": "proj2/src",
            "venv": "venv2",
            "extraPaths": ["shared_lib/src"]
        }
    ]
}
```

As usual, thanks for all your effort.

**Comment by erictraut** [comment]
We had a chance to discuss this internally. Here's what we decided:
1. I'm going to remove support for "venv" within execution environments. It has never really worked, so I don't think this will be breaking any existing projects. You can specify a single "venv" at the project level and all execution environments will inherit it.
2. I'm going to better document the fact that each file is analyzed according to the execution environment it maps to. It does not inherit an execution environment based on import chains.
3. I'll also document the fact that if you need imports to act in a transitive manner, you will need to use separate pyrightconfig.json files, one for each root. If you're using VS Code, you'll need to use the multi-root workspace feature. If you're using the CLI, you'll need to run the CLI once for each root.

**Comment by MHDante** [comment]
While I will miss the intended multi-venv feature, I'm thankful for having a defined way to set up our projects. Thank you for your work.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.120, which I just published. It will also be included in the next release of pylance.
